### ISSUE TITLE
Deadlock during shutdown of interactive window

### ISSUE THREAD

**Issue by jasonmalinowski** [dotnet/roslyn#35292]
**Version Used**: VisualStudio.16.Preview/16.1.0-pre.2.0+28822.285

**Steps to Reproduce**: Happened a few times when closing VS with the C# interactive window open.

**Expected Behavior**: It to not hang.

Dump file: `\\mlangfs1\public\jasonmal\Dumps\Roslyn\35292`. Note: there are two dump files of InteractiveHost.exe; one is the one associated with the hung process, one is associated with the devenv.exe that I was using to dump the hung processes. I dumped both to make sure I got the right one but didn't do any work to figure out which is which.

Call stack of devenv.exe:

```
>	WindowsBase.dll!System.Windows.Threading.DispatcherSynchronizationContext.Wait(System.IntPtr[] waitHandles, bool waitAll, int millisecondsTimeout)	Unknown
 	mscorlib.dll!System.Threading.SynchronizationContext.InvokeWaitMethodHelper(System.Threading.SynchronizationContext syncContext, System.IntPtr[] waitHandles, bool waitAll, int millisecondsTimeout)	Unknown
 	Microsoft.CodeAnalysis.InteractiveHost.dll!Microsoft.CodeAnalysis.Interactive.InteractiveHost.SetOutput(System.IO.TextWriter value)	Unknown
 	Microsoft.CodeAnalysis.InteractiveHost.dll!Microsoft.CodeAnalysis.Interactive.InteractiveHost.Dispose()	Unknown
 	Microsoft.CodeAnalysis.EditorFeatures.Wpf.dll!Microsoft.CodeAnalysis.Editor.Interactive.InteractiveEvaluator.Dispose()	Unknown
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Interactive.VsInteractiveWindowProvider.Create.__closeEventDelegate|1(object sender, System.EventArgs e)	Unknown
```

**Comment by jasonmalinowski** [comment]
FYI to @jinujoseph and @tmat. I think I hit this once and then also hit it maybe a second time after taking the dump. Did we change anything here recently? I didn't think so...

**Comment by tmat** [comment]
Recently no. @ivanbasov fixed output synchronization issues on shutdown a while ago.

**Comment by jasonmalinowski** [comment]
Something must not be fixed then. 😦

**Comment by ivanbasov** [comment]
This is a new deadlock. The one we fixed before happened in `InteractiveHost.RemoteService`. 

The current one is between `InteractiveHost` and `InteractiveWindow`. Both have own lock guards and both can call each other. In the situation, each of them set own lock and called another one (different threads and different methods: `Close/Dispose` vs `Write`). Each called method waited for another host to be unlocked. I enjoyed investigating the dump. Thank you, @jasonmalinowski !
